feat(collaborators): listen to member changes

anderssonjohan opened the pull request:
~~part of~~ fixes #251

Add listeners for repository collaborator events:

- member
- team

Synchronizes settings when users or teams are added, removed, and when permissions for a member is changed.

Also includes minor fix to index.test.js to make the tests run.

## Tasks

- [x] add new listeners with tests
- [x] remove excess members

anderssonjohan committed:
fix(ConfigManager): add missing log property

anderssonjohan committed:
feat(collaborators): listen to member changes

Add listeners for repository collaborator events:

- member
- team

Synchronizes settings when users or teams are added, removed, and when
permissions for a member is changed.

anderssonjohan commented:
ping @decyjphr @martinm82

decyjphr commented:
@anderssonjohan will work on it tomorrow. Thanks for the ping.

martinm82 reviewed (approved):
LGTM

martinm82 commented:
Should we update as well the README?

martinm82 commented on lib/configManager.js:
```diff
@@ -1,9 +1,10 @@
+  constructor (context, ref, log) {
     this.context = context
     this.ref = ref
+    this.log = log
```
Wouldn't it be possible to use the logger through the `context`?
```suggestion
    this.log = context.log
```

anderssonjohan replied:
Ah, great! I wasn't aware of that. I added this because of null references for `this.log` when trying to run the tests in index.test.js. I'll push a change.

anderssonjohan replied:
Also, good catch, I obviously did not update all the places where ConfigManager were created in index.js! Fixup commit pushed (will rebase when review is done)

anderssonjohan commented:
I just updated the PR description because these changes currently do not fix #251. It syncs memberships defined as policy, but the collaborators plugin does currently not remove any manually added members (users/teams). Thus, this PR should not auto-close the related issue until that part is added.

anderssonjohan commented:
@decyjphr @martinm82 I found the issue with members not being removed. In my config I have only teams and no collaborators 😅 . So for starters, I need to add `collaborators: []` to the config to enable the collaborators plugin so the code that can remove excess members is called (Collaborators.remove via Collaborators.sync). But with an empty array we never get `hasChanges = true` from `compareDeep`, because the loop over `source` has zero iterations when it's empty.
Thus, to remove manually added collaborators you need to configure a collaborator to enter that loop.

I'll add a test that verifies that `compareDeep` returns `hasChanges: true` when source (config object) is empty but target (entries from GitHub response) is not.

Example debugging session where I added manually added my github account to the repo with config set to `collaborators: []`. As you can see, the loop won't run.

<img width="900" alt="Screenshot 2023-01-24 at 21 51 00" src="https://user-images.githubusercontent.com/435885/214412005-826dffe6-ffa9-4b5e-99a5-94298bc832eb.png">

This also means that users can enable this behavior (removal of manually added collaborators) by activating the plugin. If the want the collaborators to stay there is an option like in my previous config where I didn't use that plugin at all.

anderssonjohan committed:
fix(mergedeep): hasChanges for empty source and non-empty target

Fix for empty collaborators config not removing excess members.
Using config `collaborators: []` this fix makes sure that manually
added collaborators are removed.

anderssonjohan commented:
@decyjphr @martinm82 Now the PR is complete in terms of fixing #251. 

A small note on the fix for compareDeep: I first made a version where the source and target were swapped if the source was empty. In that way the recursive call would loop through the target instead and produce the same diff as in a case where GitHub "is empty" and config is not. That, however, made one test fail for compareDeep so my commit only ensures that `hasChanges` is `true` (additions and modifications are both empty) so the code that removes the unwanted collaborators is reached.

What we had before:

```
compareDeep({x:1}, {}) => { hasChanges: true, additions: [ {x:1} ], modifications: [] }
compareDeep({}, {x:1}) => { hasChanges: false, additions: [], modifications: [] }
```

What we have now is:

```
compareDeep({x:1}, {}) => { hasChanges: true, additions: [ {x:1} ], modifications: [] }
compareDeep({}, {x:1}) => { hasChanges: true, additions: [], modifications: [] }
```

I even started to look at lodash `differenceWith` to see if it could be used with a comparator function that would compare by value and username/name attributes, but that's a much larger change that I didn't dare to look into 😅 
I hope you find this small change to `compareDeep` reasonable.

anderssonjohan commented:
I think my fixes here for empty list of collaborators also fixes the problem described by #301

decyjphr reviewed (approved):
Thanks. 🚢
